File: Maze_BFS.cpp
//Time Complexity-O(m*n)-->'m' is number of rows and 'n' is number of columns
//Space Complexity-O(m*n)-->size of queue in worst case
//Did the code execute on Leetcode? Yes

class Solution {
public:
    bool hasPath(vector<vector<int>>& maze, vector<int>& start, vector<int>& destination)     {
        vector<vector<int>>dir={{0,1},{0,-1},{1,0},{-1,0}};
        queue<vector<int>>m;
        m.push(start);
        while(!m.empty())
        {
            vector<int>k=m.front();
            m.pop();
            if(k[0]==destination[0] && k[1]==destination[1])
            {
                return true;
            }
            maze[k[0]][k[1]]=2;
            for(int i=0;i<dir.size();i++)
            {
                int row=k[0];
                int col=k[1];
                while(row>=0 && row<maze.size()&& col>=0 && col<maze[0].size()&&(maze[row][col]==0 || maze[row][col]==2))
                {
                    row+=dir[i][0];
                    col+=dir[i][1];
                }
                row-=dir[i][0];
                col-=dir[i][1];
                if(maze[row][col]!=2)
                {
                    m.push({row,col});
                }
            }
        }
        return false;
    }
    
};

File: mazeBFS.py
#Time: O(m * n * k)
#Space: O(m * n)
#Program ran on leetcode successfully

class Solution: 
    def hasPath(self, maze: List[List[int]], start: List[int], destination: List[int]):
        m = len(maze)
        n = len(maze[0])

        if maze is None or len(maze) == 0:
            return False
        
        dirs = [[-1,0], [0,1], [1,0], [0,-1]]
        q = []
        q.append(start)
        maze[start[0]][start[1]] = 2
        while q:
            curr = q[0]
            for dir in dirs:
                i = curr[0]
                j = curr[1]
                while (i >= 0 and j >= 0 and i < m and j < n and maze[i][j] != 1):
                    i += dir[0]
                    j += dir[1]
                
                i += -dir[0]
                j += -dir[1]

                if maze[i][j] != 2:
                    if i == destination[0] and j == destination[1]:
                        return True
                    q.append([i, j])
                    maze[i][j] = 2
        
        return False

